src/mbgl/map/tile_worker.cpp

```diff
@@ -32,34 +34,58 @@ TileWorker::~TileWorker() {
     style.glyphAtlas->removeGlyphs(reinterpret_cast<uintptr_t>(this));
 }
 
-Bucket* TileWorker::getBucket(const StyleLayer& layer) const {
-    std::lock_guard<std::mutex> lock(bucketsMutex);
+TileParseResult TileWorker::parseAllLayers(const GeometryTile& geometryTile) {
+    // We're doing a fresh parse of the tile, because the underlying data has changed.
+    pending.clear();
 
-    const auto it = buckets.find(layer.bucket->name);
-    if (it == buckets.end()) {
-        return nullptr;
+    for (auto i = layers.rbegin(); i != layers.rend(); i++) {
+        parseLayer(**i, geometryTile);
     }
 
-    assert(it->second);
-    return it->second.get();
+    result.state = pending.empty() ? TileData::State::parsed : TileData::State::partial;
+    return std::move(result);
 }
 
-TileParseResult TileWorker::parse(const GeometryTile& geometryTile) {
-    partialParse = false;
+TileParseResult TileWorker::parsePendingLayers() {
+    // Try parsing the remaining layers that we couldn't parse in the first step due to missing
+    // dependencies.
+    for (auto it = pending.begin(); it != pending.end();) {
+        auto& styleBucket = it->first;
+        auto& bucket = it->second;
+        assert(bucket);
+
+        if (styleBucket.type == StyleLayerType::Symbol) {
+            auto symbolBucket = dynamic_cast<SymbolBucket*>(bucket.get());
+            if (!symbolBucket->needsDependencies(*style.glyphStore, *style.sprite)) {
+                symbolBucket->addFeatures(reinterpret_cast<uintptr_t>(this), *style.spriteAtlas,
+                                          *style.glyphAtlas, *style.glyphStore, *collisionTile);
+                insertBucket(styleBucket.name, std::move(bucket));
+                pending.erase(it++);
+                continue;
+            }
+        }
 
-    for (auto i = layers.rbegin(); i != layers.rend(); i++) {
-        parseLayer(**i, geometryTile);
+        // Advance the iterator here; we're skipping this when erasing an element from this list.
+        ++it;
     }
 
-    return partialParse ? TileData::State::partial : TileData::State::parsed;
+    result.state = pending.empty() ? TileData::State::parsed : TileData::State::partial;
+    return std::move(result);
 }
 
-void TileWorker::redoPlacement(float angle, float pitch, bool collisionDebug) {
+void TileWorker::redoPlacement(
+    const std::unordered_map<std::string, std::unique_ptr<Bucket>>* buckets,
```

Comment:
That makes sense, and I couldn't think of any data races given how the worker request cancellation mechanism works. One thing that would make the reasoning a little bit clearer would be to adjust `VectorTileData::parse` so that it first explicitly does `workRequest.reset()`, and then does `workRequest = worker.parseVectorTile(...)`. (Right now, it's possible for another work task from the tile to continue to run briefly while the `parseVectorTile` request kicks off, but before the assignment to `workRequest` results in the previous task being cancelled.)